src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.Snapshotter.cs

```diff
@@ -0,0 +1,250 @@
+            private readonly PooledDictionary<BoundNode, Snapshot> _snapshotMap;
+            private readonly bool _ownsMap;
+
+            internal Snapshotter(NullableWalker walker, bool takeIncrementalSnapshot)
+            {
+                _walkerToSnapshot = walker;
+                if (takeIncrementalSnapshot)
+                {
+                    _snapshotMap = PooledDictionary<BoundNode, Snapshot>.GetInstance();
+                    _ownsMap = true;
+                }
+            }
+
+            private Snapshotter(NullableWalker newWalker, PooledDictionary<BoundNode, Snapshot> snapshotMap)
+            {
+                _walkerToSnapshot = newWalker;
+                _snapshotMap = snapshotMap;
+                _ownsMap = false;
+            }
+
+            internal void Free()
+            {
+                if (_ownsMap)
+                {
+                    _snapshotMap?.Free();
+                }
+            }
+
+            internal Snapshotter CreateForChildWalker(NullableWalker childWalker) =>
+                new Snapshotter(childWalker, _snapshotMap);
+
+            internal VariableState GetVariableState(Optional<LocalState> localState)
+            {
+                // We cache the last used dictionaries for _variableSlot and variableBySlot,
+                // in order to minimize the number of dictionaries created. As we don't ever
+                // remove or change things in these dictionaries, only add to them, we can just
+                // track whether the underlying data structures have had their sizes changed since
+                // the last time we captured them.
+                // NullableWalker.variableBySlot is allocated in an exponential fashion: every time
+                // a new slot is created that's off the end of the array, the array size is doubled.
+                // Therefore, we only want to take the allocated identifiers, and verify that any slots
+                // after the ones that we have are empty in Debug to verify. Slot 0 is always allocated,
+                // so if _lastVariableBySlot is empty then we've never cached anything and need to cache
+                if (_walkerToSnapshot._variableSlot.Count != _lastVariableSlot.Count
+                    || _lastVariableBySlot.Length == 0)
+                {
+                    var variableBySlotBuilder = ArrayBuilder<VariableIdentifier>.GetInstance(_walkerToSnapshot.variableBySlot.Length);
+                    var walkerVariableBySlot = _walkerToSnapshot.variableBySlot;
+                    int i;
+                    for (i = 0; i < walkerVariableBySlot.Length; i++)
+                    {
+                        // The first slot always does not exist for the purposes of caching and can be ignored
+                        // in the exit condition check
+                        if (i != 0 && !walkerVariableBySlot[i].Exists)
+                        {
+                            break;
```

Comment:
>break; [](start = 28, length = 6)

Can we use ```nextVariableSlot``` instead to figure out required capacity of the builder and to know how many items to copy with AddRange?